### PR TITLE
delay compsensated gradient descent

### PR DESCRIPTION
**Pull request by cstur4**
This feature is important as discussed in [issue#8744](https://github.com/tensorflow/tensorflow/issues/8744).

### CODE REVIEW

**Comment by cstur4** [comment]
I signed it!

File: tensorflow/contrib/opt/BUILD [hunk 1]
```diff
@@ -18,6 +18,7 @@ py_library(
         "python/training/adamax.py",
         "python/training/addsign.py",
         "python/training/agn_optimizer.py",
+		    "python/training/delay_compensated_gradient_descent.py",
```
Comment (alextp) [inline tensorflow/contrib/opt/BUILD]: Please run buildifier to format this file

File: tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    var_scaled_g_values = lr * (grad + lambda_ * grad * grad * (var_slice - shadow_slice))
+
+    var_t = scatter_add(var, indices, -var_scaled_g_values)
+	
+    with ops.control_dependencies([var_t]):
+      shadow_t = state_ops.assign(shadow, var_t)
+   
+    return control_flow_ops.group(*[var_t, shadow_t])
+
+  def _apply_sparse(self, grad, var):
```
Comment (alextp) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: You can just use state_ops.scatter_add for both resource and non-resource cases

File: tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        var.handle,
+        math_ops.cast(self._learning_rate_tensor, grad.dtype.base_dtype),
+        grad,
+        math_ops.cast(self._lambda_tensor, grad.dtype.base_dtype),
+        shadow.handle,
+        use_locking=self._use_locking)
+
+  def _apply_sparse_shared(self, grad, var, indices, scatter_add):
+   
+    shadow = self.get_slot(var, "shadow_{0}".format(self.worker_index))
```
Comment (alextp) [inline tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: Should the shadow variables be checkpointed? Should they be stored in a parameter server or in the local worker?

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py (hunk 3 above)
Comment (cstur4) [reply tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py]: Shadow variables should be checkpointed because the corresponding parameters are checkpointed and the checkpointed shadow variables are more reasonable than the initialization values. Shadow variables are stored in the parameter server because workers have no idea about the updating parameters in the parameter server.

**Comment by cstur4** [comment]
Hi @alextp any more advice?

**Comment by alextp** [comment]
@cstur4 tf.contrib is currently being deprecated so I don't want to add any more code to it until the replacement addons package is ready. Please see https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md for more information.

So I'll close this :-/ sorry
